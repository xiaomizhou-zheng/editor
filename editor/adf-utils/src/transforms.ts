export { transformMediaLinkMarks } from './transforms/media-link-transform';
export { transformTextLinkCodeMarks } from './transforms/text-link-code-transform';
export { transformDedupeMarks } from './transforms/dedupe-marks-transform';
export { transformNodesMissingContent } from './transforms/nodes-missing-content-transform';
export { transformIndentationMarks } from './transforms/indentation-marks-transform';
