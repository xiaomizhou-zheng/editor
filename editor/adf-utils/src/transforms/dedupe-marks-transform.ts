import { traverse } from '../traverse/traverse';
import { ADFEntity, ADFEntityMark } from '../types';

// This is the set of marks that we wont allow in duplicate to
// exist on a given node, regardless of their attributes. We do
// not include annotations here, because we do allow duplicate
// annotations as long as they have unique id attributes (valid scenario)
const markDuplicatesDisallowed = new Set([
  'strong',
  'underline',
  'textColor',
  'link',
  'em',
  'subsup',
  'strike',
]);

const maybeHasDisallowedDuplicateMarks = (node: ADFEntity): boolean => {
  const markTypes = node.marks
    ?.map((mark) => mark.type)
    ?.filter(
      (markType) =>
        // For annotations, we are performing the same cheap check
        // for duplicates by type, without considering IDs, to quickly determine
        // whether there may be potential deduping targets.
        // In the actual deduping logic in maybeRemoveDisallowedDuplicateMarks,
        // we correctly/safely dedupe annotations by their unique IDs.
        markDuplicatesDisallowed.has(markType) || markType === 'annotation',
    );

  if (!markTypes?.length) {
    return false;
  }

  return new Set(markTypes).size !== markTypes.length;
};

interface MaybeRemoveDisallowedDuplicateMarksResult {
  node?: ADFEntity;
  discardedMarks: ADFEntityMark[];
}

const maybeRemoveDisallowedDuplicateMarks = (
  node: ADFEntity,
): MaybeRemoveDisallowedDuplicateMarksResult => {
  const quota = new Map<string, boolean>();
  const annotationsQuota = new Map<string | undefined, boolean>();
  const discardedMarks: ADFEntityMark[] = [];

  markDuplicatesDisallowed.forEach((mark) => {
    quota.set(mark, false);
  });

  if (!node.marks) {
    return { discardedMarks };
  }

  let dedupedMarks = node.marks.filter((mark) => {
    const markType = mark.type;

    if (markType === 'annotation') {
      const id = mark.attrs?.id;
      if (annotationsQuota.has(id)) {
        discardedMarks.push(mark);
        return false;
      } else {
        annotationsQuota.set(id, true);
        return true;
      }
    }

    if (quota.has(markType)) {
      if (!quota.get(markType)) {
        quota.set(markType, true);
        return true;
      } else {
        discardedMarks.push(mark);
        return false;
      }
    }
    return true;
  });

  return {
    node: {
      ...node,
      marks: dedupedMarks,
    },
    discardedMarks,
  };
};

interface TransformDedupeMarksResult {
  transformedAdf: ADFEntity;
  isTransformed: boolean;
  discardedMarks: ADFEntityMark[];
}

export const transformDedupeMarks = (
  adf: ADFEntity,
): TransformDedupeMarksResult => {
  let isTransformed: boolean = false;
  let discardedMarks: ADFEntityMark[] = [];

  const transformedAdf = traverse(adf, {
    text: (node) => {
      if (maybeHasDisallowedDuplicateMarks(node)) {
        const result = maybeRemoveDisallowedDuplicateMarks(node);
        discardedMarks = result.discardedMarks;
        if (discardedMarks.length) {
          isTransformed = true;
          return result.node;
        }
      }
    },
  }) as ADFEntity;

  return {
    transformedAdf,
    isTransformed,
    discardedMarks,
  };
};
