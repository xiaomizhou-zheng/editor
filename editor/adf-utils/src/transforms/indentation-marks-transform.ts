import { traverse } from '../traverse/traverse';
import { ADFEntity } from '../types';

const hasChildHeadingWithIndentation = (node: ADFEntity): boolean =>
  node.content?.some(
    (childNode) =>
      childNode?.type === 'heading' &&
      childNode?.marks?.some((mark) => mark.type === 'indentation'),
  ) ?? false;

const removeIndentationFromHeadings = (node: ADFEntity) => {
  return {
    ...node,
    content: node.content?.map((childNode) => {
      if (childNode?.type === 'heading') {
        return {
          ...childNode,
          marks: childNode.marks?.filter((mark) => mark.type !== 'indentation'),
        };
      }
      return childNode;
    }),
  };
};

interface TransformIndentationMarksResult {
  transformedAdf: ADFEntity;
  isTransformed: boolean;
}

export const transformIndentationMarks = (
  adf: ADFEntity,
): TransformIndentationMarksResult => {
  let isTransformed: boolean = false;

  const transformedAdf = traverse(adf, {
    tableCell: (node) => {
      if (hasChildHeadingWithIndentation(node)) {
        isTransformed = true;
        return removeIndentationFromHeadings(node);
      }
    },
  }) as ADFEntity;

  return {
    transformedAdf,
    isTransformed,
  };
};
