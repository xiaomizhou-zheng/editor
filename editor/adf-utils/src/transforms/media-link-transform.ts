import { traverse } from '../traverse/traverse';
import { ADFEntity, ADFEntityMark } from '../types';

// link mark on mediaSingle is deprecated, need to move link mark to child media node
// https://product-fabric.atlassian.net/browse/ED-14043
export const transformMediaLinkMarks = (adf: ADFEntity) => {
  let isTransformed: boolean = false;
  const transformedAdf = traverse(adf, {
    mediaSingle: (node) => {
      if (!node.marks || !node.content || node.content[0]?.type !== 'media') {
        return node;
      }

      let linkMark: ADFEntityMark | null = null;
      node.marks.forEach((mark, i) => {
        if (mark.type === 'link') {
          linkMark = mark;
          node.marks?.splice(i, 1);
          isTransformed = true;
        }
      });

      if (node.marks.length === 0) {
        delete node.marks;
      }

      if (linkMark) {
        const mediaNode = node.content[0];
        // only add link mark if media node doesnt already have one
        if (mediaNode.marks?.every((mark) => mark.type !== 'link')) {
          mediaNode.marks.push(linkMark);
        } else if (!mediaNode.marks) {
          mediaNode.marks = [linkMark];
        }
      }

      return node;
    },
  });

  return {
    transformedAdf,
    isTransformed,
  };
};
