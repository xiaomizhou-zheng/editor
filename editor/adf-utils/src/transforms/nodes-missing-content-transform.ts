import { traverse } from '../traverse/traverse';
import { ADFEntity } from '../types';
import { tableRow, tableCell, paragraph, listItem } from '../builders';
import type { TableCellDefinition, TextDefinition } from '@atlaskit/adf-schema';

const getMaxColumnsCountForTable = (tableNode: ADFEntity): number => {
  let colsInRow = 1;
  tableNode.content?.forEach((childNode) => {
    if (
      childNode?.type === 'tableRow' &&
      typeof childNode.content?.length === 'number' &&
      childNode.content.length > colsInRow
    ) {
      colsInRow = childNode.content.length;
    }
  });
  return colsInRow;
};

const createValidEmptyContent = (node: ADFEntity): ADFEntity[] => {
  switch (node.type) {
    case 'tableCell':
      return [paragraph()];
    default:
      return [];
  }
};

const isEmpty = (node: ADFEntity) => node.content?.length === 0;

const fixIfTableCellInvalidEmpty = (reportTransform: ReportTransform) => (
  node: ADFEntity,
) => {
  if (isEmpty(node)) {
    reportTransform();
    return {
      ...node,
      content: createValidEmptyContent(node),
    };
  }
};

const hasNonListItemChildren = (node: ADFEntity) =>
  node.content?.some((node) => node?.type !== 'listItem');

const hasEmptyListItemChildren = (node: ADFEntity) =>
  node.content?.some(
    (childNode) => childNode?.type === 'listItem' && !childNode.content?.length,
  );

const tryCreateValidListItemWrappedChildren = (parentListNode: ADFEntity) => {
  return parentListNode.content?.map((childNode) => {
    if (childNode) {
      switch (childNode.type) {
        case 'listItem': {
          if (isEmpty(childNode)) {
            const result = listItem([paragraph()]);
            return result;
          }
          return childNode;
        }
        case 'text':
          return listItem([paragraph(childNode as TextDefinition)]);
        default:
          return listItem([childNode as any]);
      }
    }
    return childNode;
  });
};

const fixIfListParentWithInvalidListItemChildren = (
  reportTransform: ReportTransform,
) => (node: ADFEntity) => {
  if (hasNonListItemChildren(node) || hasEmptyListItemChildren(node)) {
    reportTransform();
    return {
      ...node,
      content: tryCreateValidListItemWrappedChildren(node),
    };
  }
};

const hasNonTableRowChildren = (node: ADFEntity) =>
  node.content?.some((node) => node?.type !== 'tableRow');

const tryCreateValidTableRowWrappedChildren = (parentTableNode: ADFEntity) => {
  const maxColsCount = getMaxColumnsCountForTable(parentTableNode);

  return parentTableNode.content?.map((childNode) => {
    if (childNode) {
      switch (childNode.type) {
        case 'text': {
          return tableRow([
            tableCell({})(paragraph(childNode as TextDefinition)),
            ...new Array(maxColsCount - 1).fill(tableCell({})(paragraph())),
          ]);
        }
        case 'tableCell': {
          return tableRow([childNode as TableCellDefinition]);
        }
        case 'tableRow': {
          if (isEmpty(childNode)) {
            return tableRow([
              ...new Array(maxColsCount).fill(tableCell({})(paragraph())),
            ]);
          }
          return childNode;
        }
        default:
          return childNode;
      }
    }
    return childNode;
  });
};

const hasEmptyTableRowChildren = (node: ADFEntity) =>
  node?.content?.some(
    (node) => node?.type === 'tableRow' && node?.content?.length === 0,
  );

const fixIfTableParentWithInvalidTableRowChildren = (
  reportTransform: ReportTransform,
) => (node: ADFEntity) => {
  if (hasEmptyTableRowChildren(node) || hasNonTableRowChildren(node)) {
    reportTransform();
    return {
      ...node,
      content: tryCreateValidTableRowWrappedChildren(node),
    };
  }
};

type ReportTransform = () => void;

interface TransformNodesMissingContentResult {
  transformedAdf: ADFEntity;
  isTransformed: boolean;
}

export const transformNodesMissingContent = (
  adf: ADFEntity,
): TransformNodesMissingContentResult => {
  let isTransformed: boolean = false;

  const reportTransform: ReportTransform = () => {
    isTransformed = true;
  };

  let transformedAdf = traverse(adf, {
    tableCell: fixIfTableCellInvalidEmpty(reportTransform),
  });

  transformedAdf = traverse(transformedAdf || adf, {
    bulletList: fixIfListParentWithInvalidListItemChildren(reportTransform),
    orderedList: fixIfListParentWithInvalidListItemChildren(reportTransform),
    table: fixIfTableParentWithInvalidTableRowChildren(reportTransform),
  }) as ADFEntity;

  return {
    transformedAdf,
    isTransformed,
  };
};
