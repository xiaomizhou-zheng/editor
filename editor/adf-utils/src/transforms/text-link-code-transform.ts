import { traverse } from '../traverse/traverse';
import { ADFEntity } from '../types';

const hasLinkAndCodeMarks = (node: ADFEntity) => {
  const marks = node.marks?.map((mark) => mark.type);
  return marks?.includes('link') && marks?.includes('code');
};

const removeCodeMarks = (node: ADFEntity) => {
  if (node.marks) {
    return {
      ...node,
      marks: node.marks.filter((mark) => mark.type !== 'code'),
    };
  }
  return node;
};

// See: HOT-97965 https://product-fabric.atlassian.net/browse/ED-14400
// We declared in code mark spec that links and marks should not co-exist on
// text nodes. This util strips code marks from bad text nodes and preserves links.
export const transformTextLinkCodeMarks = (adf: ADFEntity) => {
  let isTransformed: boolean = false;
  const transformedAdf = traverse(adf, {
    text: (node) => {
      if (hasLinkAndCodeMarks(node)) {
        isTransformed = true;
        return removeCodeMarks(node);
      }
    },
  });

  return {
    transformedAdf,
    isTransformed,
  };
};
