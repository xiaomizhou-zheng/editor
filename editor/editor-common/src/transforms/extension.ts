import { Schema, Slice } from 'prosemirror-model';

import { mapFragment } from '../utils/slice';

/**
 * Lift content out of "open" top-level bodiedExtensions.
 * Will not work if bodiedExtensions are nested, or when bodiedExtensions are not in the top level
 */
export const transformSliceToRemoveOpenBodiedExtension = (
  slice: Slice,
  schema: Schema,
) => {
  const { bodiedExtension } = schema.nodes;

  const fragment = mapFragment(slice.content, (node, parent, index) => {
    if (node.type === bodiedExtension && !parent) {
      const currentNodeIsAtStartAndIsOpen = slice.openStart && index === 0;
      const currentNodeIsAtEndAndIsOpen =
        slice.openEnd && index + 1 === slice.content.childCount;

      if (currentNodeIsAtStartAndIsOpen || currentNodeIsAtEndAndIsOpen) {
        return node.content;
      }
    }
    return node;
  });

  // If the first/last child has changed - then we know we've removed a bodied extension & to decrement the open depth
  return new Slice(
    fragment,
    fragment.firstChild &&
    fragment.firstChild!.type !== slice.content.firstChild!.type
      ? slice.openStart - 1
      : slice.openStart,
    fragment.lastChild &&
    fragment.lastChild!.type !== slice.content.lastChild!.type
      ? slice.openEnd - 1
      : slice.openEnd,
  );
};
