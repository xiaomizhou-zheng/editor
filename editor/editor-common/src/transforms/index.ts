export {
  unwrapContentFromLayout,
  removeLayoutFromFirstChild,
  removeLayoutFromLastChild,
  transformSliceToRemoveOpenLayoutNodes,
} from './layout';
export {
  findExpand,
  transformSliceToRemoveOpenExpand,
  transformSliceNestedExpandToExpand,
} from './expand';
export { transformSliceToRemoveOpenBodiedExtension } from './extension';
