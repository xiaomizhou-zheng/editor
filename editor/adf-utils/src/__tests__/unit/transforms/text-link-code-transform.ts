import { transformTextLinkCodeMarks } from '../../../transforms/text-link-code-transform';

import textLinkCodeAdf from './__fixtures__/text-link-code-adf.json';
import textValidMarksAdf from './__fixtures__/text-valid-marks-adf.json';

describe('transformTextLinkCodeMarks', () => {
  it('should remove code marks and preserve link marks on text node', () => {
    let { isTransformed, transformedAdf } = transformTextLinkCodeMarks(
      textLinkCodeAdf,
    );
    expect(isTransformed).toEqual(true);
    expect(transformedAdf).toMatchSnapshot();
  });

  it('should not change text nodes with valid mark combinations', () => {
    let { isTransformed, transformedAdf } = transformTextLinkCodeMarks(
      textValidMarksAdf,
    );
    expect(isTransformed).toEqual(false);
    expect(transformedAdf).toMatchSnapshot();
  });
});
