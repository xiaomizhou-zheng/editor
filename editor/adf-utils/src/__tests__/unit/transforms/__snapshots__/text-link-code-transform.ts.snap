// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transformTextLinkCodeMarks should not change text nodes with valid mark combinations 1`] = `
Object {
  "content": Array [
    Object {
      "content": Array [
        Object {
          "marks": Array [
            Object {
              "attrs": Object {
                "href": "http://google.com",
              },
              "type": "link",
            },
            Object {
              "attrs": Object {
                "color": "#6554c0",
              },
              "type": "textColor",
            },
          ],
          "text": "world",
          "type": "text",
        },
        Object {
          "marks": Array [
            Object {
              "attrs": Object {
                "color": "#6554c0",
              },
              "type": "textColor",
            },
          ],
          "text": " ",
          "type": "text",
        },
        Object {
          "marks": Array [
            Object {
              "type": "em",
            },
            Object {
              "type": "strong",
            },
            Object {
              "attrs": Object {
                "color": "#6554c0",
              },
              "type": "textColor",
            },
            Object {
              "type": "strike",
            },
            Object {
              "attrs": Object {
                "type": "sup",
              },
              "type": "subsup",
            },
          ],
          "text": "more",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "doc",
  "version": 1,
}
`;

exports[`transformTextLinkCodeMarks should remove code marks and preserve link marks on text node 1`] = `
Object {
  "content": Array [
    Object {
      "content": Array [
        Object {
          "marks": Array [
            Object {
              "attrs": Object {
                "href": "https://www.atlassian.com",
              },
              "type": "link",
            },
          ],
          "text": "text with link and code mark",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "doc",
  "version": 1,
}
`;
